### PR TITLE
fixes #20422; prepareStrMutation for toOpenArray (extend to more types)

### LINKED ISSUE #20422: Mutating a var parameter through a mutable view triggers SIGSEGV
### What happened?

```nim
{.experimental: "views".}

proc f(a: var string) =
  var v = a.toOpenArray(1, 3)
  v[0] = 'a'  # <--

var a = "Hello"
f(a)
```
```
Traceback (most recent call last)
D:\User\test.nim(116)   test
D:\User\test.nim(113)   f
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
Error: execution of an external program failed: 'D:\User\test.exe'
```

### Nim Version

1.7.1

### Current Standard Output Logs

```shell
Traceback (most recent call last)
D:\User\test.nim(116)   test
D:\User\test.nim(113)   f
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
Error: execution of an external program failed: 'D:\User\test.exe'
```

### Expected Standard Output Logs

_No response_

### Possible Solution

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by ringabout**
`toOpenArray` always return `tyOpenArray` type. `prepareStrMutation` is not called in this case:

```nim
block:
  proc f(a: var string) =
    var v = a.toOpenArray(1, 3)
    v[0] = 'a'  # <--
  var a = "Hello"
  f(a)
```
`a` is a string literal, so we need to call `prepareStrMutation` before modification. 

However, we shouldn't call `prepareStrMutation` for immutable strings, otherwise it causes segfault.

```nim
block:
  proc f(a: string) =
    var v = a.toOpenArray(1, 3)
    doAssert v[0] == 'H'
  f("Hello")
```

In order to differentiate these two cases, only parameters from `skVar` or `skParam` with a `tyVar` type
are considered.

### CODE REVIEW

**Comment by Araq** [comment]
I don't agree with this patch. In general people should know about `prepareMutation` when they fiddle with it via openArray.

**Comment by ringabout** [comment]
Yeah, then we should document it for toOpenArray and close the issue.

**Comment by arnetheduck** [comment]
> In general people should know about `prepareMutation` when they fiddle with it via openArray.

@Araq this cripples `var openArray[char]` to the point that it's practically unusable - generally, it's a good idea to write API that take the minimum type they need so that all functions that don't need a "full" `string` don't take it - however, if I know that the caller of my api _also_ has to call an unrelated magic function before they use it, it no longer makes sense. - ie "knowning about copy-on-write" is difficult to teach because it is only later, at runtime, that you *sometimes* find out that you haven't done it - read the manual is not an answer here - every single function that exposes `var openArray` would have to document this weird gotcha separately which is unfeasible, and equally wasteful as having to call `prepmutation` manually.

basically, to make this sane for consumers of any functions, for every function I have to write two functions: one that takes `var oa[char]` and another boilerplate that takes `var string` then calls `prep` and my `oa` function.

**Comment by Araq** [comment]
We emit `prepareMutation` automatically in other places and it produced performance regressions. Now I err on the other side and make it explicit and it's wrong again. Sometimes you just cannot win.
